### PR TITLE
[NFR]: T13856 jwt support

### PR DESCRIPTION
**Pull request by niden**
Hello!

* Type: new feature
* Link to issue: #13856 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [ ] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Added JWT support
- Token
- Parser
- Builder
- Validator
- Signers (None, HMAC)
- Added Base64 encode/decodeUrl helper class

This is fully covered by tests but would like it to be tested in an application so please do not merge yet.

Thanks

### CODE REVIEW

**Comment by sergeyklay** [comment]
Could you please send this to 4..1.x branch

**Comment by ruudboon** [comment]
Awesome work

**Comment by niden** [comment]
@sergeyklay We dont have a 4.1.x branch. Should I create one?

**Comment by ruudboon** [comment]
> @sergeyklay We dont have a 4.1.x branch. Should I create one?

Sounds good to me. And let's merge the fixes from 4.0.x from time to time to make sure we're not too far behind.

**Comment by niden** [comment]
I have one more method to add to verify the signature and this should be ready.

File: phalcon/Http/JWT/Token/Token.zep [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * @property Signature $signature
+ *
+ * @link https://tools.ietf.org/html/rfc7519
+ */
+class Token
+{
+    /**
+     * @var Item
+     */
+    private claims;
```
Comment (sergeyklay) [inline phalcon/Http/JWT/Token/Token.zep]:
```suggestion
    private claims { get };
```

File: phalcon/Http/JWT/Token/Token.zep [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * @var Item
+     */
+    private jose;
```
Comment (sergeyklay) [inline phalcon/Http/JWT/Token/Token.zep]:
```suggestion
    private headers { get };
```

File: phalcon/Http/JWT/Token/Token.zep [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * @var Signature
+     */
+    private signature;
```
Comment (sergeyklay) [inline phalcon/Http/JWT/Token/Token.zep]:
```suggestion
    private signature { get };
```

File: phalcon/Http/JWT/Signer/AbstractSigner.zep [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * Class AbstractSigner
+ *
+ * @property string $algo
+ */
+abstract class AbstractSigner implements SignerInterface
+{
+    /**
+     * @var string
+     */
+    protected algo;
```
Comment (sergeyklay) [inline phalcon/Http/JWT/Signer/AbstractSigner.zep]:
```suggestion
    protected algorithm { get };
```

**Comment by sergeyklay** [comment]
@niden Consider proposed changes

File: phalcon/Http/JWT/Token/Token.zep [hunk 5]
```diff
@@ -24,17 +24,17 @@ class Token
...
      * @var Item
      */
-    private claims;
+    private claims { get };
 
     /**
      * @var Item
      */
-    private jose;
+    private jose { get };
```
Comment (sergeyklay) [inline phalcon/Http/JWT/Token/Token.zep]: @niden Did you merged `getJose` ?

Comment (niden) [reply phalcon/Http/JWT/Token/Token.zep]: Its already fixed bud.

Comment (sergeyklay) [reply phalcon/Http/JWT/Token/Token.zep]: Thank you

**Pull request merged**